### PR TITLE
[JENKINS-61478] Style improvements to alert banners

### PR DESCRIPTION
**Pull request by romenrg**
See [JENKINS-61478](https://issues.jenkins-ci.org/browse/JENKINS-61478).

Recent changes have caused the alert banners to look broken (their height no loner matching the navbar height). This was a good excuse to try to improve them, so apart from  fixing the height in my first commit, I also experimented a bit, copying the styles from [bootstrap alerts](https://getbootstrap.com/docs/4.0/components/alerts/#examples), removing the opacity and switching to material icons. The fade-in and fade-out effects remain, as well as the duration of the alert.

### Screenshots

In the Jira ticket, the steps to reproduce the "success" and "copied" (default) alerts were described. In the following screenshots you can see the "before" and "after" for each case.

#### Success alert ("apply")

Before:
before-success-alert

After (using "alert-success" styles from bootstrap and the check_circle icon from material icons):
after-success-alert

#### Default alert ("copied")

The use case for copying an API token, described in the Jira ticket, corresponds to the default alert.

Before:
before-default-alert

After (using the "alert-info" styles from boostrap):
after-default-alert

#### Warning and Error alerts

Apart from the two cases described above and reported in the ticket, I've noticed there are also "warning" and "error" alerts. I was not able to reproduce a behavior that will trigger warning or error alerts. However, I did modify the styles and tested them by temporarily replacing the "success" alert with these other two.

Before:
after-warning-alert
before-error-alert

After (using the corresponding bootstrap styles and material icon "help_outline" for warning, and "highlight_off" for error):
before-warning-alert
after-error-alert

### Proposed changelog entries

* Style improvements on alert banners

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] JIRA issue is well described
- [X] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [X] Appropriate autotests or explanation to why this change has no tests
- [X] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@fqueiruga
@daniel-beck
@josephbrueggen
@uhafner
@batmat 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

### CODE REVIEW

**Review by fqueiruga — APPROVED** [review]

**Review by daniel-beck — APPROVED** [review]
Works well. Thanks for taking care of this!

The "Success" banner is a bit washed out, making it difficult to distinguish (except for the message) from the default banner, but that may only be because I'm used to the existing style.

**Comment by romenrg** [comment]
@fqueiruga Thanks for the feedback. I agree ideally styles should be added to CSS classes. If I have some time I can try to extract it. However, I tried to make as little changes as possible to the existing script, which was setting styles in the script itself, that's why I kept it that way.

@daniel-beck Thanks Daniel. The success color is indeed less intense now than it used to be. Personally, I like it as it is now. That said, I'm open to suggestions. I just used the bootstrap color considering it some sort of a standard.

**Review by res0nance — APPROVED** [review]
Looks much better now.
Are these new colours friendly to those who are colour blind?

**Review by timja — APPROVED** [review]

File: war/src/main/webapp/scripts/hudson-behavior.js
Comment (uhafner) [reply war/src/main/webapp/scripts/hudson-behavior.js]: 👍 Using a `class` would also help to keep the colors from our new color palette in sync for all items that should use the same semantic. (And theming will be much simpler as well.)

**Comment by uhafner** [comment]
Good work, thanks, it looks much better now!

I must admit that I am still not a fan of the material icons, but that is a matter of preference. (And as maintainer of the Font Awesome icons API plugin I'm somewhat biased).

Shouldn't we pick different icon elements for the styles? Here the example from the react-material home page:

image

They use an exclamation mark in a triangle for warnings (This type I am also using for the warnings plugin) and an exclamation mark for errors in a circle. I think a question mark is something more appropriate for a help icon.

**Comment by romenrg** [comment]
Thanks @uhafner for the feedback! I love your suggestions for the icons!

I'll make that change, copying the icons (and its styles) used by the [react-material plugin](https://material-ui.com/components/alert/#simple-alerts) (I'm adding this as a note for myself, to keep the link handy)

**Comment by uhafner** [comment]
I'm not sure if we should copy icons from somewhere else (@fqueiruga)? Doesn't the bundled set of material  icons contain similar icons? I think we should start using icons from *one* icon set, if possible.

**Comment by fqueiruga** [comment]
Indeed, we should continue adding icons from the official material sprite set. Agree with you Ulli on that.

**Comment by fqueiruga** [comment]
Anyway, iconography enhancements are coming, probably in one month or so. One of the things we will have is a coherent icon strategy, so I wouldn't be really worried about this single case ATM.

**Comment by romenrg** [comment]
@fqueiruga and @uhafner my intention was to say that I will copy which icon they used for each alert; as well as the colors used for icons. I was not referring to copying new icons. To me they look exactly the same (or very very similar) to the material icons we are already using). And since those we are including as SVG, we can modify their color without adding any new icons.

**Comment by daniel-beck** [comment]
Are these internal changes? If not, could you please update screenshots and re-request reviews?

**Comment by romenrg** [comment]
@daniel-beck Sure, I was about to add the new screenshots after the commits, but took a small break. You were quick! :)

Here are the new screenshots after my latest changes:
latest-success-alert
latest-default-alert
latest-warning-alert
latest-error-alert

I have implemented @uhafner's suggestions of styles and icons. I've used the closest icons I found from the material icons that were already included in Jenkins.

I have also addressed @fqueiruga 's suggestions of making icons bigger and, more importantly, moving the alert styles and animations to CSS (less). It was an interesting challenge, but I'm quite happy with the result. This is a significant improvement in maintenance, I would say.

So, the resulting styles are the ones of the screenshots included in this comment. The resulting animations are quite similar to the ones we had originally. I'm doing a fade-in / fade-out effect, but now CSS native :)

Also pinging @josephbrueggen, who was waiting for these latest changes to review.

That's all from my update. Looking forward to your reviews. Hope everyone likes these improvements!

File: war/src/main/less/base/style.less
Comment (fqueiruga) [inline war/src/main/less/base/style.less]:
```suggestion
    display: block;
    will-change: opacity;
```

File: war/src/main/less/base/style.less
Comment (fqueiruga) [inline war/src/main/less/base/style.less]:
```suggestion
  -webkit-animation: fadein 350ms ease-out 1 normal forwards;
```

File: war/src/main/less/base/style.less
Comment (fqueiruga) [inline war/src/main/less/base/style.less]:
```suggestion
  -webkit-animation: fadeout 350ms ease-in 1 normal forwards;
```

**Review by fqueiruga — CHANGES_REQUESTED** [review]
Changes are looking great. I have added some suggestions tuning the animation durations and easing functions.

Also, I see a lack of vendor prefixes for the `animation` shorthand, only with `-webkit` and no unprefixed declaration. Did you test it in multiple browsers?

**Comment by timja** [comment]
> Also, I see a lack of vendor prefixes for the `animation` shorthand, only with `-webkit` and no unprefixed declaration. Did you test it in multiple browsers?

doesn't look like its needed?
https://caniuse.com/#feat=css-animation

**Review by daniel-beck — APPROVED** [review]
Very nice

**Comment by romenrg** [comment]
I've just added a new commit addressing @fqueiruga 's latest suggestions.

Regarding the vendor prefixes, I was indeed testing changes in Chrome, planned to add them later and had forgotten. They are now included. As @timja commented above, it seems using just the non-prefixed version should be enough, but I added all the rest as well, since there should not be harm on including them all.

Hope this is ready to merge now :)

**Review by fqueiruga — APPROVED** [review]

**Comment by timja** [comment]
> I've just added a new commit addressing @fqueiruga 's latest suggestions.
> 
> Regarding the vendor prefixes, I was indeed testing changes in Chrome, planned to add them later and had forgotten. They are now included. As @timja commented above, it seems using just the non-prefixed version should be enough, but I added all the rest as well, since there should not be harm on including them all.
> 
> Hope this is ready to merge now :)

we should really add one of those plugins that handles if for you, they decide whether the prefixes should be added based on the browsers you support.

I don't think we should be adding all those prefixes, but as you say it doesn't do harm (other than possibly making it harder to maintain)

Adding ready-for-merge, thanks Romén and all the reviewers, let's merge it tomorrow if there's no negative feedback

**Comment by fqueiruga** [comment]
> we should really add one of those plugins that handles if for you, they decide whether the prefixes should be added based on the browsers you support.

There's an open ticket for doing that on the CSS revamp epic https://issues.jenkins-ci.org/browse/JENKINS-60924.

I'm leaving it available for community contributions until I have the time to work on it.

**Comment by romenrg** [comment]
BTW, I'm curious about the "lts-candidate" label. When do we use it? Should we use it in this ticket?

**Review by uhafner — APPROVED** [review]
Nice! 

(Is there a special reason that you are using the [highlight_off](https://material.io/resources/icons/?search=Highlight&icon=highlight_off&style=outline) icon for errors and not the [error_outline](https://material.io/resources/icons/?search=war&style=outline)?)

**Comment by romenrg** [comment]
@uhafner yes, the reason is that at the moment we only have the "action" and "social" packages of material icons included in Jenkins Core. The "error_outline" icon is in another package (alert), so I adapted to use the ones that were already available to avoid adding new packages.

**Comment by fqueiruga** [comment]
I don't mind having the current icon but FWIW we should be progressively adding all icon sets

**Review by oleg-nenashev — APPROVED** [review]
Tested it on my machine, works great!

**Comment by daniel-beck** [comment]
Caused [JENKINS-62065](https://issues.jenkins-ci.org/browse/JENKINS-62065).

**Pull request merged**
